src/Stratis.Bitcoin.Features.SmartContracts/Consensus/Rules/AllowedScriptTypesRule.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Threading.Tasks;
+using NBitcoin;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.MemoryPool;
+using Stratis.SmartContracts;
+using Stratis.SmartContracts.Core;
+using Block = NBitcoin.Block;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.Consensus.Rules
+{
+    /// <summary>
+    /// Enforces that only certain script types are used on the network.
+    /// </summary>
+    public class AllowedScriptTypesRule : PartialValidationConsensusRule, ISmartContractMempoolRule
+    {
+        protected ISmartContractCoinviewRule ContractCoinviewRule { get; private set; }
+
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            base.Initialize();
+            this.ContractCoinviewRule = (ISmartContractCoinviewRule)this.Parent;
+        }
+
+        public override Task RunAsync(RuleContext context)
+        {
+            Block block = context.ValidationContext.BlockToValidate;
+
+            foreach (Transaction transaction in block.Transactions)
+            {
+                CheckTransaction(transaction);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public void CheckTransaction(MempoolValidationContext context)
+        {
+            CheckTransaction(context.Transaction);
+        }
+
+        private void CheckTransaction(Transaction transaction)
+        {
+            // Why dodge coinbase?
+            // 1) Coinbase can only be written by Authority nodes anyhow.
+            // 2) Coinbase inputs look weird, are tough to validate.
+            if (!transaction.IsCoinBase)
+            {
+                foreach (TxOut output in transaction.Outputs)
+                {
+                    CheckOutput(output);
+                }
+
+                foreach (TxIn input in transaction.Inputs)
+                {
+                    CheckInput(input);
+                }
+            }
+        }
+
+        private void CheckOutput(TxOut output)
+        {
+            if (output.ScriptPubKey.IsSmartContractExec())
+                return;
+
+            if (output.ScriptPubKey.IsSmartContractInternalCall())
+                return;
+
+            if (PayToPubkeyHashTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
+                return;
+
+            if (PayToScriptHashTemplate.Instance.CheckScriptPubKey(output.ScriptPubKey))
```

Comment:
I am not sure this is what you meant, shouldn't it be `PayToMultiSigTemplate` for non-P2SH (i.e. raw) multisig?

If you do it this way then arbitrary P2SH scriptPubKeys will be allowed through.

Reply:
Basically they are two different script structures:

1. Native multisig where `OP_CHECKMULTISIG` appears in a (long) scriptPubKey containing the pubkeys of all the multisig participants. The spending scriptSig presents the signatures of `m` of the `n` participants. This is space-inefficient for receiving funds as every 'incoming' transaction has all the keys in it. Spending takes a moderate amount of space.

2. P2SH multisig, where the scriptPubKey only contains the hash of the redeem script. This is considerably more space-efficient for receiving funds, and facilitates arbitrary redeem scripts. Spending therefore takes a variable amount of space depending on the complexity of the redeem script.

Basically we can use either for sidechains, it's just that we are currently using P2SH as it is usually the 'preferred' methodology (as the `3xxxxxx` addresses that result are easy to work with from a user perspective). How would you determine the sender for either variant, in any case? P2SH in particular might make it impossible to do so in a meaningful way.

Reply:
Oooh you raise a really important point - if we do allow a non-P2PKH, then we need some rules about when they can be used because we need the 'sender'

I believe the most reasonable is: If calling a smart contract, the first input to the transaction MUST be P2PKH.

Reply:
Is the multisig / p2sh implementation likely to change @zeptin ?

If so, it may be worth delaying that part of this rule until it's finalised

Reply:
Well, we would only change it if you need it to be changed, I think.

I need to go over how you call smart contracts in your implementation. A user could try to get the federation to send funds to a smart contract destination address, which seems like it would cause problems.